### PR TITLE
Fixed triple-reporting lint issues

### LINKED ISSUE #1336: GitHub Actions is triple-reporting lint issues
We do GitHub Actions builds on multiple platforms, and each one reports the same errors inline in our PRs.  This is annoying, and it would be better to only have lint issues reported once:

Screen Shot 2020-11-12 at 11 49 40 AM

### PR DESCRIPTION
**Pull request by yzwdroid**
## Issue This PR Addresses
Fixes #1336

## Type of Change

- [x] **Bugfix**: Change which fixes an issue
- [ ] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
only run `test-ci` when the os is `ubuntu`

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Review by humphd — CHANGES_REQUESTED** [review]
The problem I see with this approach is that we no longer run tests on the other platforms, in which case there is no point having them listed in the matrix.

@raygervais do you have any insight into other ways to reduce the lint spam without shutting off the build matrix?

**Comment by humphd** [comment]
@yzwdroid this needs a rebase, and an update on where we stand.  Are you hoping to get this done in time for 0.3?  If so, we need to finalize and merge this today.

**Comment by yzwdroid** [comment]
I am thinking if we have a better solution for this. I already have a PR merged for release 0.3.  Thanks.
I'll do the rebase now.

**Review by humphd — CHANGES_REQUESTED** [review]
Alright, I did some research and here's a better way, let's separate the lint and test scripts, and do something like [Svelte is doing](https://github.com/sveltejs/svelte/actions/runs/375353097/workflow).  Here is their workflow file:

```yml
name: CI
on: [push, pull_request]
jobs:
  Tests:
    runs-on: ${{ matrix.os }}
    timeout-minutes: 10
    strategy:
      matrix:
        node-version: [8, 10, 12, 14]
        os: [ubuntu-latest, windows-latest, macOS-latest]
    steps:
    - uses: actions/checkout@v1
    - uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}
    - run: npm install
    - run: npm test
      env:
        CI: true
  Lint:
    runs-on: ubuntu-latest
    timeout-minutes: 2
    steps:
    - uses: actions/checkout@v1
    - uses: actions/setup-node@v1
    - run: 'npm i && npm run lint'
  Unit:
    runs-on: ${{ matrix.os }}
    timeout-minutes: 5
    strategy:
      matrix:
        os: [ubuntu-latest, windows-latest, macOS-latest]
    steps:
    - uses: actions/checkout@v1
    - uses: actions/setup-node@v1
    - run: 'npm i && npm run test:unit'
```

To do this, we'll need to get rid of [`test-ci`](https://github.com/Seneca-CDOT/telescope/blob/master/package.json#L48) and do these three things separately:

1. `npm run prettier-check` to make sure the PR has proper formatting.  Do this only on linux
1. `npm run eslint` to make sure the PR passes lint. Do this only on linux
1. `npm run test` to make sure the PR passes tests.  Do this on all OSes

**Comment by yzwdroid** [comment]
Thanks for the help @humphd . I changed the condition as you mentioned.

**Review by humphd — CHANGES_REQUESTED** [review]
You haven't done what I asked.  Can you split up the jobs so that we have separate prettier, eslint, and test phases that can run in parallel?

**Comment by yzwdroid** [comment]
I am a little confused about this approach. If we run all the jobs async, then we need to do `npm install` three times in the ubuntu server. Is that slow down a little bit?

**Comment by humphd** [comment]
Not if we also add caching, see https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows.

**Comment by yzwdroid** [comment]
I'll add the cache, and commit it later.

File: .github/workflows/node-js-ci.yml [hunk 1]
```diff
@@ -28,4 +28,16 @@ jobs:
       - run: npm install
       - run: cp env.example .env
       - run: npm run build
```
Comment (humphd) [inline .github/workflows/node-js-ci.yml:30]: Let's get rid of this build step too, since we aren't using the build in CI.

**Review by humphd — CHANGES_REQUESTED** [review]

File: .github/workflows/node-js-ci.yml (hunk 1 above)
Comment (humphd) [reply .github/workflows/node-js-ci.yml:30]: See https://github.com/Seneca-CDOT/telescope/pull/1444/commits/b615e5c12d143c6138f0c4bb2bd27d82355a6eea, where I'm doing that for another PR too.

**Comment by yzwdroid** [comment]
After I add the cache, the build time reduced from 3m 30 s to  2m 54s on Ubuntu from my repo [https://github.com/yzwdroid/telescope/runs/1449300001?check_suite_focus=true](https://github.com/yzwdroid/telescope/runs/1449300001?check_suite_focus=true). I'm not sure if I have done it right.

File: .github/workflows/node-js-ci.yml [hunk 2]
```diff
@@ -19,13 +19,36 @@ jobs:
...
     steps:
       - uses: actions/checkout@v2
+      - name: Cache node modules
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-node-modules
+        with:
+          # npm cache files are stored in `~/.npm` on Linux/macOS
+          path: ~/.npm
+          key: ${{ runner.os }}-build-${{ env.cache-name }}-${{ hashFiles('**/package-lock.json') }}
```
Comment (humphd) [inline .github/workflows/node-js-ci.yml:29]: We don't use `package-lock.json` in most cases.  What does `${{ hashFiles('**/package-lock.json') }}` do and can we live without it?

File: .github/workflows/node-js-ci.yml [hunk 3]
```diff
@@ -19,13 +19,36 @@ jobs:
...
         uses: actions/setup-node@v1
         with:
           node-version: ${{ matrix.node-version }}
 
       - run: npm install
       - run: cp env.example .env
-      - run: npm run build
-      - run: npm run test-ci
+      - run: npm run test
+  Prettier:
```
Comment (humphd) [inline .github/workflows/node-js-ci.yml:43]: `Prettier-Check`

**Review by humphd — CHANGES_REQUESTED** [review]

File: .github/workflows/node-js-ci.yml (hunk 2 above)
Comment (yzwdroid) [reply .github/workflows/node-js-ci.yml:29]: we recache the node modules when this file changed.

Comment (humphd) [reply .github/workflows/node-js-ci.yml:29]: Right, but when we don't have it?

**Comment by humphd** [comment]
Let's get this rebased for the conflict that's in node-js-ci.yml.

**Comment by yzwdroid** [comment]
I rebased the current commit, Is there any code to be changed?

**Review by humphd — APPROVED** [review]

**Review by manekenpix — APPROVED** [review]

**Pull request merged**